File: Cora/Objective-C/Managers/TaskManager.m
#import "TaskManager.h"


@implementation TaskManager


+ (TaskManager *)sharedInstance {

	static TaskManager *sharedInstance = nil;
	static dispatch_once_t onceToken;

	dispatch_once(&onceToken, ^{

		sharedInstance = [self new];

	});

	return sharedInstance;

}


- (void)launchDarwinTask {

	[[UserDefaultsManager sharedInstance] loadSwitchState];

	if(![UserDefaultsManager sharedInstance].switchState) return;

	NSTask *task = [NSTask new];
	[task setLaunchPath:@"/bin/sh"];
	[task setArguments:@[@"-c", @"uname -a"]];

	NSPipe *outputPipe = [NSPipe pipe];
	task.standardOutput = outputPipe;
	[task launchAndReturnError:nil];

	NSData *outputData = [[outputPipe fileHandleForReading] readDataToEndOfFile];
	self.darwinString = [[NSString alloc] initWithData:outputData encoding:NSUTF8StringEncoding];

}


- (void)launchUptimeTask {

	[[UserDefaultsManager sharedInstance] loadSwitchState];

	NSTask *task = [NSTask new];
	[task setLaunchPath:@"/bin/sh"];
	[task setArguments:@[@"-c", @"uptime"]];

	NSPipe *outputPipe = [NSPipe pipe];
	task.standardOutput = outputPipe;
	[task launchAndReturnError:nil];

	NSData *outputData = [[outputPipe fileHandleForReading] readDataToEndOfFile];
	self.uptimeString = [[NSString alloc] initWithData:outputData encoding:NSUTF8StringEncoding];

}


@end

File: Cora/Swift/Managers/TaskManager.swift
import Foundation


final class TaskManager {
	
	static let sharedInstance = TaskManager()
	var darwinString: String?
	var uptimeString: String?
	
	private init() {}
	
	func launchDarwinTask() {

		UserDefaultsManager.sharedInstance.loadSwitchState()

		guard let switchState = UserDefaultsManager.sharedInstance.switchState else {
			return
		}

		if !switchState {
			return
		}

		let darwinTask = NSTask()
		darwinTask.arguments = ["-c", "uname -a"]
		darwinTask.launchPath = "/bin/sh"
		
		let pipe = Pipe()
		darwinTask.standardOutput = pipe
		darwinTask.launch()
		
		let data = pipe.fileHandleForReading.readDataToEndOfFile()
		darwinString = String(data: data, encoding: .utf8)

	}

	func launchUptimeTask () {

		let uptimeTask = NSTask()
		uptimeTask.arguments = ["-c", "uptime"]
		uptimeTask.launchPath = "/bin/sh"
		
		let pipe = Pipe()
		uptimeTask.standardOutput = pipe
		uptimeTask.launch()
		
		let data = pipe.fileHandleForReading.readDataToEndOfFile()
		uptimeString = String(data: data, encoding: .utf8)

	}

}
